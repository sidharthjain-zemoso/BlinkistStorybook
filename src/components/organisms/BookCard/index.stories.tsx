import { ComponentStory } from "@storybook/react";
import BookCard from "./index";

export default {
  title: "Organisms/BookCard",
  component: BookCard,
  argTypes: {
    variant: {
      control: {
        type: "select",
        options: [
          "defaultCard",
          "addToLibCard",
          "finishedCard",
          "readAgainCard",
        ],
      },
    },
  },
};

const Template: ComponentStory<typeof BookCard> = (args) => (
  <BookCard {...args} />
);

export const Default = Template.bind({});
Default.args = {
  variant: "defaultCard",
  title: "Bring Your Human to Work",
  author: "Erica Keswin",
  time: "13",
  reads: "5k",
  imageSrc: "1",
};
