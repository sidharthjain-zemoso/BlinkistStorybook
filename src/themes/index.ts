import { createTheme } from "@mui/material/styles";

declare module "@mui/material/Typography" {
  interface TypographyPropsVariantOverrides {
    heading: true;
    heading3: true;
    subtitle3: true;
    body3: true;
    caption1: true;
    caption2: true;
    caption3: true;
  }
}

declare module "@mui/material/styles" {
  interface TypographyVariants {
    heading: React.CSSProperties;
    heading3: React.CSSProperties;
    subtitle3: React.CSSProperties;
    body3: React.CSSProperties;
    caption1: React.CSSProperties;
    caption2: React.CSSProperties;
    caption3: React.CSSProperties;
  }

  interface TypographyVariantsOptions {
    heading?: React.CSSProperties;
    heading3: React.CSSProperties;
    subtitle3?: React.CSSProperties;
    body3?: React.CSSProperties;
    caption1?: React.CSSProperties;
    caption2?: React.CSSProperties;
    caption3: React.CSSProperties;
  }

  interface PaletteColor {
    darker?: string;
    border?: string;
    placeholder?: string;
  }

  interface SimplePaletteColorOptions {
    darker?: string;
    placeholder?: string;
    border?: string;
  }

  interface Palette {
    greyscale: Palette["primary"];
    textcolor: Palette["primary"];
    backgroundcolor: Palette["primary"];
  }

  interface PaletteOptions {
    greyscale?: PaletteOptions["primary"];
    textcolor?: PaletteOptions["primary"];
    backgroundcolor?: PaletteOptions["primary"];
  }
}

export const theme = createTheme({
  components: {
    MuiButtonBase: {
      styleOverrides: {
        root: {
          // "&:hover": {
          //   backgroundColor: "none",
          // },
        },
      },
    },
  },
  spacing: 4,
  palette: {
    primary: {
      light: "#E2F2E9",
      main: "#2CE080",
      dark: "#22C870",
      darker: "#20BA68",
    },
    secondary: {
      light: "#DFE8F6",
      main: "#0365F2",
    },
    greyscale: {
      light: "#BAC9CF",
      main: "#6C787F",
      dark: "#3A4649",
      darker: "#042330",
    },
    textcolor: {
      light: "#6D787E",
      main: "#03314B",
      placeholder: "#747575",
      dark: "#042330",
    },
    backgroundcolor: {
      light: "#F7F7F7",
      main: "#F1F6F4",
    },
  },
  typography: {
    heading: {
      fontSize: "36px",
      lineHeight: "45px",
      // fontWeight: 700,
      fontFamily: "Cera Pro Medium",
    },
    heading3: {
      fontSize: "24px",
      lineHeight: "30.17px",
      // fontWeight: 700,
      fontFamily: "Cera Pro Medium",
    },
    subtitle1: {
      fontSize: "24px",
      lineHeight: "32px",
      // fontWeight: 500,
      fontFamily: "Cera Pro Medium",
    },
    subtitle2: {
      fontSize: "18px",
      lineHeight: "24px",
      fontWeight: 700,
      fontFamily: "Cera Pro Light",
    },
    subtitle3: {
      fontSize: "18px",
      lineHeight: "24px",
      // fontWeight: 400,
      fontFamily: "Cera Pro Light",
    },
    body1: {
      fontSize: "16px",
      lineHeight: "20px",
      // fontWeight: 700,
      fontFamily: "Cera Pro Medium",
    },
    body2: {
      fontSize: "16px",
      lineHeight: "24px",
      // fontWeight: 400,
      fontFamily: "Cera Pro Light",
    },
    body3: {
      fontSize: "22px",
      lineHeight: "20px",
      // fontWeight: 400,
      fontFamily: "Cera Pro Light",
    },
    caption1: {
      fontSize: "14px",
      lineHeight: "22px",
      // fontWeight: 400,
      fontFamily: "Cera Pro Light",
    },
    caption2: {
      fontSize: "14px",
      lineHeight: "22px",
      // fontWeight: 700,
      fontFamily: "Cera Pro Medium",
    },
    caption3: {
      fontSize: "14px",
      lineHeight: "17.6px",
      // fontWeight: 400,
      fontFamily: "Cera Pro Light",
    },
  },
});

// theme.overrides = {
//   ...theme.overrides,
//   MuiButton: {
//     root: {
//     }
//   }
// }
