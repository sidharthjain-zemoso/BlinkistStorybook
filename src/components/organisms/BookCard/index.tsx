import Card from "@mui/material/Card";
import CardMedia from "@mui/material/CardMedia";
import CardContent from "@mui/material/CardContent";
import TypographyText from "../../atoms/Typography";
import AddIcon from "@mui/icons-material/Add";
import { theme } from "../../../themes";
import Icon from "../../atoms/Icon";
import CustomButton from "../../atoms/Button";
import { makeStyles } from "@mui/styles";
import ProgressBar from "../../atoms/ProgressBar";

const useStyles = makeStyles({
  cardContainer: {
    border: "1px solid #E1ECFC",
    borderRadius: "8px",
  },
  cardContentWrapperELement: {
    maxWidth: "350px",
    padding: "5px",
    marginTop: "1.5px",
  },
  readElement: {
    display: "flex",
    justifyContent: "space-between",
  },
});

interface BookCardProps {
  variant: "defaultCard" | "addToLibCard" | "finishedCard" | "readAgainCard";
  title: string;
  author: string;
  time: string;
  reads?: string;
  imageSrc: string;
  className?: string;
  buttonOnClick?: React.MouseEventHandler<HTMLElement>;
}

export function BookCard(props: BookCardProps) {
  const styleClasses = useStyles();
  const {
    variant,
    className,
    buttonOnClick,
    title,
    author,
    time,
    reads,
    imageSrc,
  } = props;
  return (
    <Card
      data-testid="book-card-id"
      className={`${className} ${styleClasses.cardContainer}`}
      sx={{ maxWidth: 345, height: 550 }}
    >
      <CardMedia
        component="img"
        height="350"
        image={`/res/cover images/${imageSrc}.png`}
        alt="Book Cover"
      />
      <CardContent>
        <div className="card-content-wrapper">
          <div className={styleClasses.cardContentWrapperELement}>
            <TypographyText
              variant="subtitle1"
              component="div"
              color={theme.palette.textcolor.main}
            >
              {title}
            </TypographyText>
          </div>
          <div className={styleClasses.cardContentWrapperELement}>
            <TypographyText
              variant="body1"
              component="div"
              color={theme.palette.textcolor.light}
            >
              {author}
            </TypographyText>
          </div>
          <div
            className={`${styleClasses.readElement} ${styleClasses.cardContentWrapperELement}`}
          >
            <div
              className="read-time"
              style={{
                display: "flex",
                alignItems: "center",
              }}
            >
              <Icon url="/res/icons/clock.png" />
              <TypographyText
                component="div"
                variant="caption3"
                color={theme.palette.textcolor.light}
              >
                {time}-minute read
              </TypographyText>
            </div>
            {reads ? (
              <div
                className="read-num"
                style={{
                  display: "flex",
                  alignItems: "center",
                }}
              >
                <Icon url="/res/icons/user.png" />
                <TypographyText
                  component="div"
                  variant="caption3"
                  color={theme.palette.textcolor.light}
                >
                  {reads} reads
                </TypographyText>
              </div>
            ) : (
              ""
            )}
          </div>
          {variant === "defaultCard" ? (
            <div className={styleClasses.cardContentWrapperELement}>
              <Icon
                url="/res/icons/more.png"
                sx={{
                  marginLeft: "auto",
                  display: "flex",
                  marginTop: "10px",
                }}
              />
            </div>
          ) : variant === "finishedCard" || variant === "readAgainCard" ? (
            <div style={{ textAlign: "center", marginTop: "20px" }}>
              <TypographyText variant="body1" component="div" color="secondary">
                {variant === "finishedCard" ? "Finished" : "Read Again"}
              </TypographyText>
            </div>
          ) : (
            ""
          )}
        </div>
      </CardContent>
      {variant === "defaultCard" ||
      variant === "finishedCard" ||
      variant === "readAgainCard" ? (
        <ProgressBar
          sx={{ marginTop: variant === "defaultCard" ? "20px" : "6px" }}
          progress={variant === "readAgainCard" ? 100 : 30}
        />
      ) : variant === "addToLibCard" ? (
        <CustomButton
          id="cardButtonId"
          startIcon={<AddIcon />}
          className="addToLib"
          size="large"
          variant="outlined"
          sx={{ width: "100%", marginTop: "10px" }}
          onclick={buttonOnClick}
        >
          <TypographyText variant="body1" component="div">
            Add To Library
          </TypographyText>
        </CustomButton>
      ) : (
        ""
      )}
    </Card>
  );
}

export default BookCard;
