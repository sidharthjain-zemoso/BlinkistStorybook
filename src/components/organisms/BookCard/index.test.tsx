import { fireEvent, render, screen } from "@testing-library/react";
import BookCard from "./";
import { ThemeProvider } from "@mui/material/styles";
import { theme } from "../../../themes";

describe("Book Card", () => {
  test("renders a default BookCard", () => {
    render(
      <ThemeProvider theme={theme}>
        <BookCard
          variant="defaultCard"
          title="Bring Your Human to Work"
          author="Erica Keswin"
          time="13"
          reads="5k"
          imageSrc="1"
        />
      </ThemeProvider>
    );

    const bookCard = screen.getByTestId("book-card-id");
    const authorElement = screen.getByText(/bring your human to work/i);
    const titleElement = screen.getByText(/erica keswin/i);
    // const imageElement = screen.getByRole("img");
    const timeElement = screen.getByText(/13-minute read/i);
    const readsElement = screen.getByText(/5k reads/i);

    expect(authorElement).toBeInTheDocument();
    expect(titleElement).toBeInTheDocument();
    // expect(imageElement).toBeInTheDocument();
    expect(timeElement).toBeInTheDocument();
    expect(readsElement).toBeInTheDocument();
    expect(bookCard).toBeInTheDocument();

    expect(bookCard).toMatchSnapshot();
  });

  test("renders a addToLib variant BookCard", () => {
    render(
      <ThemeProvider theme={theme}>
        <BookCard
          variant="addToLibCard"
          title="Bring Your Human to Work"
          author="Erica Keswin"
          time="13"
          reads="5k"
          imageSrc="1"
        />
      </ThemeProvider>
    );

    const bookCard = screen.getByTestId("book-card-id");
    const authorElement = screen.getByText(/bring your human to work/i);
    const titleElement = screen.getByText(/erica keswin/i);
    // const imageElement = screen.getByRole("img");
    const timeElement = screen.getByText(/13-minute read/i);
    const readsElement = screen.getByText(/5k reads/i);
    const addToLibButton = screen.getByRole("button", {
      name: "Add To Library",
    });

    expect(authorElement).toBeInTheDocument();
    expect(titleElement).toBeInTheDocument();
    // expect(imageElement).toBeInTheDocument();
    expect(timeElement).toBeInTheDocument();
    expect(readsElement).toBeInTheDocument();
    expect(addToLibButton).toBeInTheDocument();
    expect(bookCard).toBeInTheDocument();

    expect(bookCard).toMatchSnapshot();
  });
});
