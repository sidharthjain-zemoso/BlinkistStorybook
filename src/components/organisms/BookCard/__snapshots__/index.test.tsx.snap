// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Card renders a addToLib variant BookCard 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root undefined makeStyles-cardContainer-4 css-31xq66-MuiPaper-root-MuiCard-root"
  data-testid="book-card-id"
>
  <img
    alt="Book Cover"
    class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
    height="350"
    src="/res/cover images/1.png"
  />
  <div
    class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      class="card-content-wrapper"
    >
      <div
        class="makeStyles-cardContentWrapperELement-5"
      >
        <div
          class="MuiTypography-root MuiTypography-subtitle1 css-19oa8dr-MuiTypography-root"
        >
          Bring Your Human to Work
        </div>
      </div>
      <div
        class="makeStyles-cardContentWrapperELement-5"
      >
        <div
          class="MuiTypography-root MuiTypography-body1 css-nzdi-MuiTypography-root"
        >
          Erica Keswin
        </div>
      </div>
      <div
        class="makeStyles-readElement-6 makeStyles-cardContentWrapperELement-5"
      >
        <div
          class="read-time"
          style="display: flex; align-items: center;"
        >
          <img
            alt="icon"
            src="/res/icons/clock.png"
            style="margin-right: 5px;"
          />
          <div
            class="MuiTypography-root MuiTypography-caption3 css-1igz8yl-MuiTypography-root"
          >
            13
            -minute read
          </div>
        </div>
        <div
          class="read-num"
          style="display: flex; align-items: center;"
        >
          <img
            alt="icon"
            src="/res/icons/user.png"
            style="margin-right: 5px;"
          />
          <div
            class="MuiTypography-root MuiTypography-caption3 css-1igz8yl-MuiTypography-root"
          >
            5k
             reads
          </div>
        </div>
      </div>
      
    </div>
  </div>
  <button
    class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButtonBase-root makeStyles-addToLib-7 css-1b1cqo3-MuiButtonBase-root-MuiButton-root"
    id="cardButtonId"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </span>
    <div
      class="MuiTypography-root MuiTypography-body1 css-1ty6goh-MuiTypography-root"
    >
      Add To Library
    </div>
  </button>
</div>
`;

exports[`Book Card renders a default BookCard 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root undefined makeStyles-cardContainer-1 css-31xq66-MuiPaper-root-MuiCard-root"
  data-testid="book-card-id"
>
  <img
    alt="Book Cover"
    class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
    height="350"
    src="/res/cover images/1.png"
  />
  <div
    class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <div
      class="card-content-wrapper"
    >
      <div
        class="makeStyles-cardContentWrapperELement-2"
      >
        <div
          class="MuiTypography-root MuiTypography-subtitle1 css-19oa8dr-MuiTypography-root"
        >
          Bring Your Human to Work
        </div>
      </div>
      <div
        class="makeStyles-cardContentWrapperELement-2"
      >
        <div
          class="MuiTypography-root MuiTypography-body1 css-nzdi-MuiTypography-root"
        >
          Erica Keswin
        </div>
      </div>
      <div
        class="makeStyles-readElement-3 makeStyles-cardContentWrapperELement-2"
      >
        <div
          class="read-time"
          style="display: flex; align-items: center;"
        >
          <img
            alt="icon"
            src="/res/icons/clock.png"
            style="margin-right: 5px;"
          />
          <div
            class="MuiTypography-root MuiTypography-caption3 css-1igz8yl-MuiTypography-root"
          >
            13
            -minute read
          </div>
        </div>
        <div
          class="read-num"
          style="display: flex; align-items: center;"
        >
          <img
            alt="icon"
            src="/res/icons/user.png"
            style="margin-right: 5px;"
          />
          <div
            class="MuiTypography-root MuiTypography-caption3 css-1igz8yl-MuiTypography-root"
          >
            5k
             reads
          </div>
        </div>
      </div>
      <div
        class="makeStyles-cardContentWrapperELement-2"
      >
        <img
          alt="icon"
          src="/res/icons/more.png"
          style="margin-right: 5px; margin-left: auto; display: flex; margin-top: 10px;"
        />
      </div>
    </div>
  </div>
  <div
    data-testid="progress-bar-id"
    style="height: 15px; background-color: rgb(241, 246, 244); margin-top: 20px;"
  >
    <div
      style="width: 30%; height: 15px; background-color: rgb(225, 236, 252);"
    />
  </div>
</div>
`;
